string.c

```diff
@@ -543,6 +518,150 @@ rb_fstring(VALUE str)
+
+    for (;;) {
+        VALUE candidate = table->entries[idx];
+
+        if (candidate == FSTRING_TABLE_EMPTY) {
+            // Not in table
+            if (insert) {
+                int prev_count = RUBY_ATOMIC_FETCH_ADD(table->count, 1);
+
+                if (UNLIKELY(prev_count > table->capacity / 2)) {
+                    fstring_try_resize();
+                    goto retry;
+                }
+
+                VALUE found = RUBY_ATOMIC_VALUE_CAS(table->entries[idx], FSTRING_TABLE_EMPTY, value);
+                if (found == FSTRING_TABLE_EMPTY) {
+                    // Success! Our value was inserted
+                    return value;
+                } else {
+                    // Nothing was inserted
+                    RUBY_ATOMIC_DEC(table->count); // we didn't end up inserting
+
+                    // Another thread won the race, try again at the same location
+                    continue;
+                }
+            } else {
+
+                return Qfalse;
+            }
+        } else if (candidate == FSTRING_TABLE_TOMBSTONE) {
+            // Deleted entry, continue searching
+        } else if (!fstring_cmp(candidate, value)) {
+            // We've found a match
+            if (rb_objspace_garbage_object_p(candidate)) {
+                // We found an object which is about to be GC'd
+                // Skip it and mark it is a tombstone
+                RUBY_ATOMIC_VALUE_CAS(table->entries[idx], candidate, FSTRING_TABLE_TOMBSTONE);
+            } else {
+                return candidate;
+            }
+        }
+
+        // linear probing (TODO: something better)
+        idx++;
+        idx %= table->capacity;
+    }
+}
+
+// Rebuilds the table
+static void fstring_try_resize(void) {
+    RB_VM_LOCK_ENTER();
+
+    // Stop all other threads to prevent concurrent reads as we resize
+    // TODO: Through extra work we should be able to perform this resize without locks or a barrier
+    rb_vm_barrier();
+
+    // This may overcount by up to the number of threads concurrently attempting to insert
+    // GC may also happen between now and the table being rebuilt
+    int expected_count = fstring_table.count - fstring_table.deleted_entries;
+
+    VALUE *old_entries = fstring_table.entries;
+    int old_capacity = fstring_table.capacity;
+    int new_capacity = old_capacity * 2;
+    if (new_capacity > expected_count * 4) {
+        new_capacity = old_capacity;
+    }
+
+    // May cause GC and therefore deletes, so must hapen first
+    VALUE *new_entries = ZALLOC_N(VALUE, new_capacity);
+
+    fstring_table.capacity = new_capacity;
+    fstring_table.entries = new_entries;
+    fstring_table.count = 0;
+    fstring_table.deleted_entries = 0;
+
+    for (int i = 0; i < old_capacity; i++) {
+        VALUE val = old_entries[i];
+        if (val == FSTRING_TABLE_EMPTY)     continue;
+        if (val == FSTRING_TABLE_TOMBSTONE) continue;
+        fstring_upsert(str_do_hash(val), val, true);
```

Comment:
Is there a reason you use `str_do_hash`? I don't know if you've seen https://github.com/ruby/ruby/pull/12881 but `str_do_hash` does cause quite a lot of collisions in the `fstring_table`.

Also most if not all fstrings have a precomputed hash. So I think it would be better if you re-used `fstring_hash`.

Reply:
Thanks! I missed that, we can use `fstring_hash`.

One thing I also haven't settled on is whether the hash table should cache the hash value for each string. It's probably safer to do so but if we thought the precomputed hash was there very often maybe we could avoid it.

I saw https://bugs.ruby-lang.org/issues/21172 when working on this and I think have an additional fix for that in this PR: in delete we compare by identity, so it shouldn't be possible for equivalent strings to be evicted.

Reply:
> if we thought the precomputed hash was there very often maybe we could avoid it.

All literal strings will have it (unless they can't be embedded), and all embeded strings that have enough space left.

Not too sure what the exact rate is in practice, but it would be easy to compute from a heap dump.

That said, in term of locality, it's an extra pointer chasing, but I don't know if that makes that much of a difference.